### ISSUE TITLE
Help with modifying ARM assembly code

### ISSUE THREAD

**Issue by jfoucher** [dp111/PicoTube#3]
Hi,
as you're aware, I am currently working on a [VIC-20 emulator for the Raspberry Pi Pico](http://forum.6502.org/viewtopic.php?f=8&t=6663). However, I am having a hard time modifying the assembly code to get it to call an external C function. Specifically, I am having trouble doing anything in the LOAD_BYTE macro. Event if I just add the following : 
```
   push {r0-r3}
   pop {r0-r3}
```
before the `ldrb  \rdst,[ rmem, \rsrc]` line, the Pico crashes immediately (in my ignorance, I have adjusted the `tablend` label to  `tablend: . = 256*(I_ALIGN+1)` to get it to fit. Could that be what's causing the issue ? )

Basically what I would like to achieve would be to replace `ldrb  \rdst,[ rmem, \rsrc]` with a call to a C function that gets passed the address and returns the data.

Any guidance to help me along while I'm familiarizing myself with ARM assembly would be great.

Thanks!

**Comment by hoglet67** [comment]
I'm sure Dominic will respond shortly, but here's my take.

The instuction table has a fixed size slot for each instruction which allows a computed branch to be used for instruction dispatch.

Your change has cause one (or more) of the instructions to overflow it's slot.

Unfortunately, changing the tablend label to tablend: . = 256*(I_ALIGN+1) will just bypass the compiler error, but won't work in practice, because the computed branch won't go to the right place.

You have a couple of options:

1. Increase the slot size, which is set with I_ALIGN_BITS in copro-65tubeasm.h

I_ALIGN_BITS is currently 6, which gives a 64 byte / 32 instruction slot size.

Increasing it to 7 will double the code size, which will probably slow things down due to cache effects. You'll quite possibly run out of RAM though, and probably hit lots of other issues.

So this is probably not the best option.

2. Figure out which instruction is overflowing and try to re-write that so it fits into the slot. 

You can use the objdump command to see the generaed code:
```
arm-none-eabi-objdump -D PicoTube.elf
```
I've just had a quick look, and it looks like it's opcodes 61 and E1 that are overflowing by one instruction.

You would need to try to rewrite these to somehow use a branch to jump to additional code in an "overflow" code area.

None of this is likely to be straighforward!

Dave

**Comment by dp111** [comment]
Just started to write my response and Dave has correctly summarised the
issue.

The code runs directly in ram so no cache issues. Changing  I_ALIGN_BITS to
7 would be a simple way to go except it overflows a ram block.  With some
linker magic I'm sure that is possible.

tablend is a bit of magic for the assembler to tell us if the
instruction table overflows

The best way would be to identify instructions that overflow and jump to
some extra code hidden at the end of a nearby NOP instruction. My
modification to BRK overflowed so I hid the extra few instructions needed
at the end of NOP ( 0x02) . So for 0xE1  I would put the SBC macro at the
end of 0xE2 ( just like 0x02). 0x61 has already overflowed so I have made
part of it overflow to 0x62. ADC6 has a first bounce parameter so you
should be able to use that. These macros are a bit complex so use objdump
on the original code to see what code you need for ADC then check your new
version has all the code required.

**Comment by jfoucher** [comment]
Wow ok thanks for the quick replies! Let me rephrase to make sure I got it right: there are some short and some long 6502 instructions, but all of them are limited to 32 ARM instructions. If they are longer than that the extra code must be “hidden” in shorter 6502 instructions such as NOPs, otherwise the instruction table is messed up and calls to 6502 instructions will go to the wrong line. Everything makes so much more sense now, thanks!
